Src/Amr/AMReX_extrapolater_2D_K.H

```diff
@@ -0,0 +1,234 @@
+            }
+         }
+      }
+      // xhi, yhi
+      {
+         int i = hi.x+1;
+         int j = hi.y+1;
+         if ( mask(i,j,k) == crsecell ) {
+            if ( ( mask(i-1,j,k) == finecell ) ||
+                 ( mask(i,j-1,k) == finecell ) ) {
+               data(i,j,k,n) = ( mask(i-1,j,k) * data(i-1,j,k,n) + mask(i,j-1,k) * data(i,j-1,k,n) )
+                               / ( mask(i-1,j,k) + mask(i,j-1,k) );
+            } else {
+               data(i,j,k,n) = data(i-1,j-1,k,n);
+            }
+         }
+      }
+      // Edges
+      // xlo, y-valid
+      {
+         int i = lo.x-1;
+         for (int j = lo.y; j <= hi.y; ++j) {
+            if ( mask(i,j,k) == crsecell ) {
+               data(i,j,k,n) = ( mask(i,j-1,k) * data(i,j-1,k,n) + mask(i+1,j,k) + mask(i,j+1,k) * data(i,j+1,k,n) )
+                               / ( mask(i,j-1,k) + 1 + mask(i,j+1,k) );
+            }
+         }
+      }
+      // xhi, y-valid
+      {
+         int i = hi.x+1;
+         for (int j = lo.y; j <= hi.y; ++j) {
+            if ( mask(i,j,k) == crsecell ) {
+               data(i,j,k,n) = ( mask(i,j-1,k) * data(i,j-1,k,n) + mask(i-1,j,k) + mask(i,j+1,k) * data(i,j+1,k,n) )
+                               / ( mask(i,j-1,k) + 1 + mask(i,j+1,k) );
+            }
+         }
+      }
+      // x-valid, ylo
+      {
+         int j = lo.y-1;
+         for (int i = lo.x; i <= hi.x; ++i) {
+            if ( mask(i,j,k) == crsecell ) {
+               data(i,j,k,n) = ( mask(i-1,j,k) * data(i-1,j,k,n) + mask(i+1,j,k) * data(i+1,j,k,n) + data(i,j+1,k,n) )
+                               / ( mask(i-1,j,k) + mask(i+1,j,k) + 1 );
+            }
+         }
+      }
+      // x-valid, yhi
+      {
+         int j = hi.y+1;
+         for (int i = lo.x; i <= hi.x; ++i) {
+            if ( mask(i,j,k) == crsecell ) {
+               data(i,j,k,n) = ( mask(i-1,j,k) * data(i-1,j,k,n) + mask(i+1,j,k) * data(i+1,j,k,n) + data(i,j-1,k,n) )
+                               / ( mask(i-1,j,k) + mask(i+1,j,k) + 1 );
+            }
+         }
+      }
+   }
+}
+
+AMREX_GPU_HOST_DEVICE
+AMREX_FORCE_INLINE
+void
+amrex_first_order_extrap_gpu(int i, int j, int k, int n,
+                             amrex::Box const& bx,
+                             amrex::Array4<const int>   const& mask,
+                             amrex::Array4<amrex::Real> const& data) noexcept
+{
+   using namespace amrex::literals;
+
+   constexpr int finecell = 1;
+   constexpr int crsecell = 0;
+
+   const auto lo = amrex::lbound(bx);
+   const auto hi = amrex::ubound(bx);
+
+   // set all crse cells to zero first
+   if (mask(i,j,k) == crsecell) data(i,j,k,n) = 0.0_rt;
+
+   // Corners
+   // xlo, ylo
+   if ( (i == lo.x-1) && (j == lo.y-1) ) {
+      if ( mask(i,j,k) == crsecell ) {
+         if ( ( mask(i+1,j,k) == finecell ) ||
+              ( mask(i,j+1,k) == finecell ) ) {
+            data(i,j,k,n) = ( mask(i+1,j,k) * data(i+1,j,k,n) + mask(i,j+1,k) * data(i,j+1,k,n) )
```

Comment:
We may have to write this as
```
if ((mask(i+1,j,k) == finecell) && (mask(i,j+1,k) == finecell)) {
} else if (mask(i+1,j,k) == finecell) {
} else if (mask(i,j+1,k) == finecell) {
} else {
}
```
The reason for that is data(i+1,j,k) or data(i,j+1,k,n) may contains nan.  In the CPU version, we have a loop in the beginning to set all coarse cells to zero.  But here we do not have that.  We should also remove the lines
```
// set all crse cells to zero first
   if (mask(i,j,k) == crsecell) data(i,j,k,n) = 0.0_rt;
```
because it won't be able to serve the purpose like in the CPU case.

Reply:
Maybe it's easier to launch two kernels as you suggested earlier, otherwise the 3d version is going to have so many branches.